Lombiq.Hosting.Tenants.Management/Controllers/ShellSettingsEditorController.cs

```diff
@@ -0,0 +1,89 @@
+using Newtonsoft.Json;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Configuration;
+using OrchardCore.Locking.Distributed;
+using OrchardCore.Modules;
+using OrchardCore.Mvc.Core.Utilities;
+using OrchardCore.Tenants.Controllers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using static OrchardCore.Tenants.Permissions;
+
+namespace Lombiq.Hosting.Tenants.Management.Controllers;
+
+[Feature(FeatureNames.ShellSettingsEditor)]
+public class ShellSettingsEditorController : Controller
+{
+    private readonly IAuthorizationService _authorizationService;
+    private readonly IShellHost _shellHost;
+    private readonly IShellConfigurationSources _shellConfigurationSources;
+    private readonly IDistributedLock _distributedLock;
+
+    public ShellSettingsEditorController(
+        IAuthorizationService authorizationService,
+        IShellHost shellHost,
+        IShellConfigurationSources shellConfigurationSources,
+        IDistributedLock distributedLock)
+    {
+        _authorizationService = authorizationService;
+        _shellHost = shellHost;
+        _shellConfigurationSources = shellConfigurationSources;
+        _distributedLock = distributedLock;
+    }
+
+    [HttpPost]
+    [ValidateAntiForgeryToken]
+    public async Task<IActionResult> Edit(ShellSettingsEditorViewModel model)
+    {
+        if (!await _authorizationService.AuthorizeAsync(User, ManageTenants) ||
+            !_shellHost.TryGetSettings(model.TenantId, out var shellSettings))
+        {
+            return NotFound();
+        }
+
+        var settingsDictionary = JsonConvert.DeserializeObject<Dictionary<string, string>>(model.Json);
+        var newSettings = new Dictionary<string, string>();
+
+        foreach (var key in settingsDictionary.Keys.Where(key => string.IsNullOrEmpty(settingsDictionary[key])))
+        {
+            settingsDictionary[key] = null;
+        }
+
+        foreach (var key in settingsDictionary.Keys)
+        {
+            if (shellSettings[key] != settingsDictionary[key])
+            {
+                var tenantSettingsPrefix = $"{model.TenantId}Prefix:{key}";
+                newSettings[tenantSettingsPrefix] = settingsDictionary[key];
+                newSettings[key] = settingsDictionary[key];
+            }
+        }
+
+        // Try to acquire a lock before using the scope, so that a next process gets the last committed data.
+        var (locker, locked) = await _distributedLock.TryAcquireLockAsync("SHELL_SETTINGS_EDITOR_LOCK", TimeSpan.MaxValue);
+        if (!locked)
+        {
+            throw new TimeoutException($"Failed to acquire a lock before saving settings to the tenant: {model.TenantId}");
+        }
+
+        await using var acquiredLock = locker;
+
+        await _shellConfigurationSources.SaveAsync(shellSettings.Name, newSettings);
+        await _shellHost.ReloadShellContextAsync(shellSettings);
+
```

Comment:
@Piedone @wAsnk 

Why not trying to use `UpdateShellSettingsAsync()`, maybe you only want to save the tenant config but not the tenant settings (to the shared `tenants.json` by default), but this is what we do when editing and then save a tenant.

By using `UpdateShellSettingsAsync()` the settings `VersionId` is updated which may be important, the same lock is used, here the lock doesn't prevent `ShellConfigurationSources` to save at the same time (see locking below).

Also keeping `ShellSettingsManager` and `ShellConfigurationSources` working together may help to check if a config entry should be updated or removed.

If there is any problem to use `UpdateShellSettingsAsync()` I can help, but maybe not possible in your use case.

---

About locking.

The default `ShellConfigurationSources` saving to a file doesn't use a lock itself, this is done by `ShellSettingsManager` and we decided to use a local lock assuming that in a distributed environment other instances don't use the same file system, personally I think we could have but Sebastien doesn't like locks, even less distributed locks ;)

`DatabaseShellConfigurationSources` doesn't use an additional lock, we rely on the database transaction and to keep data consistency we use `session.Save(..., checkConcurrency: true)` so that if there is a concurrency exception data are not persisted. About `BlobShellConfigurationSources` I don't know how it may behave.

Reply:
> Why not trying to use UpdateShellSettingsAsync(), maybe you only want to save the tenant config but not the tenant settings (to the shared tenants.json by default), but this is what we do when editing and then save a tenant.

We can use `UpdateShellSettingsAsync` instead of `ReloadShellContextAsync` to update the version and reload context also. Other than that I'm not sure how could we use `ShellSettingsManager` and `ShellConfigurationSources` to check if something should be updated. We are checking it now directly by comparing `ShellSettings` and this new feature settings.

> About locking.

@Piedone I can't add anything to this topic.

Reply:
Re locking: for Azure blobs, [last writer wins](https://learn.microsoft.com/en-us/azure/storage/blobs/concurrency-manage#:~:text=If%20you%20don%27t%20explicitly%20specify%20a%20strategy%2C%20then%20by%20default%20the%20last%20writer%20wins.). I think this is what happens in SQL as well: i.e. if we don't do any conflict resolution, then simply the most recent value will be saved.

This is actually just fine, unless we lock in a way that's surfaced on the UI too, this should be an extreme case to cause any issues; and there won't be a data consistency issue even during concurrent writes.

So, we don't actually need locking here at all. However, we do need to use `IShellSettingsManager` to save the shell settings, not the lower-level `IShellConfigurationSources`.

Thanks @jtkech!

Reply:
@Piedone  We can't use IShellSettingsManager for our use case as I pointed it out in my previous comments.

Reply:
`IShellSettingsManager.SaveSettingsAsync()` updates shell settings via `IShellsSettingsSources`. It can also be used to retrieve shell settings of a tenant. Why can't we use it?

Reply:
OK, so ideally we'd use `IShellHost.UpdateShellSettingsAsync()` but that uses `IShellSettingsManager`, and we can't use that because it doesn't support saving config keys with more than one section.

Please explain this issue, based on all of this, in a new OC issue, and add a comment in the code here linking to that so we can know why we're doing this. The rest of the code needn't change.

Reply:
Yes, I forgot this part. I may try it if I have time but would be interesting to check what happens if in `ShellSettingsManager` we just remove:

    .Where(s => !s.GetChildren().Any())